### DISCUSSION TITLE
[Suggestion] Pull Dual Pricing into the core

### DISCUSSION THREAD

**Discussion by scottcwilson** [zencart/zencart#5773] [Ideas]
This mod is incredibly useful and would represent a significant new capability for Zen Cart.

**Comment by wiztechinc** [comment]
When someone needs it, it's great, no question. I have very few carts that have used it though, definitely less than 10%

**Comment by judygunderson** [comment]
I think Admin Login as Customer would be great if it is not already included!

**Comment by scottcwilson** [comment]
Judy, I quite agree - would you please open a separate issue about that so we can keep this ticket on the Dual Pricing topic?  Thanks.

**Comment by drbyte** [comment]
Keep in mind that pulling mods into the core implicitly expects "us" to do all the support of the feature and maintain the feature. 
Are you volunteering to take on that burden for the foreseeable future?

**Comment by scottcwilson** [comment]
For as long as I can draw breath, sir.

**Comment by wiztechinc** [comment]
> For as long as I can draw breath, sir.

And we hope you will be around a long long time. But I assume you are all aware how many have left or died - in my case, I ran into major medical issues and have simply not had the time or ability to keep up with updating mods that I wrote. (hee, hee, I'm back!)  Chris has a very valid point if one is looking long term.

**Comment by scottcwilson** [comment]
The Good Lord willing and the creek don't rise, I have many releases left in me. :)

**Comment by torvista** [comment]
> not had the time or ability to keep up with updating mods that I wrote

Me neither....hence I think the more in the core the better, so "problems" are more in the public domain/in more peoples faces and more likely to get attention/fixed.
Once you have touched a Plugin/mod it sticks to you and everyone thinks its your baby/the community input disappears.

**Comment by wiztechinc** [comment]
> > not had the time or ability to keep up with updating mods that I wrote
> 
> Me neither....hence I think the more in the core the better, so "problems" are more in the public domain/in more peoples faces and more likely to get attention/fixed.
> Once you have touched a Plugin/mod it sticks to you and everyone thinks its your baby/the community input disappears.

Boy oh boy are you right! After trying to do the right thing and upload changes to a mod, everyone thinks you got the answers when you just made some minor fixes/changes. One of the chief reasons I pulled back from fixing other mods. I just keep them to myself now.

**Comment by torvista** [comment]
I don't use either yet, but absolutely need to. My sales have gradually shifted to 90% trade, and shops are now **expecting** to be forced to buy online from their distributors.
To summarise from what I have just been reading:
stock Group Discount allows a global discount based on the group the customer belongs to.
Inadequate.
mod Dual Pricing allows four price levels per product and customers can be assigned any of them.
This will become the third essential mod for me (Advanced Shipper and Products Options Stock since you asked), so yes I would vote this in the core.

**Comment by lankeeyankee** [comment]
me too, we've been using it for 10+ years. I would call it essential

**Comment by dakanji** [comment]
Rather than adding the whole kitchen sink to the core, what should happen is that some modules should be officially supported and actually, the core should be as lean as possible. 

A subtle but important difference. Still means needing to support them but this will simplify development as a whole.

**Comment by torvista** [comment]
> the core should be as lean as possible

I feel that is valid from a developer point of view.
But ZC is not for developers, it is for business-owners and in my view should have a longer list of core functionality than the competition, if it wants to compete.
There needs to be more input requested from the forum as to what should be included, most users have no need to venture into Github and won't comment here.

**Comment by lankeeyankee** [comment]
I would agree with the "officially supported" line when the new plugin system is operational. But for now, adding useful features to the core means the average user has to dig into core hacking less, which makes for faster adoption of this over other systems. Steve is spot-on, zen cart is targeted to business owners, look at Chris's tagline in his zen cart profile. It needs to be feature-rich, modern, and easy to use.

**Comment by dakanji** [comment]
I don't think the typical user cares about the mechanism by which functionality is provided as long as it is provided in an accessible manner. 

Things don't have to be in the core for ZC to be feature-rich, modern, and easy to use

**Comment by scottcwilson** [comment]
Guys, please remember that this mod makes a giant set of changes to includes/functions/functions_prices, which is why I'd like to integrate it into the core so this work doesn't have to get done every release.

**Comment by lankeeyankee** [comment]
exactly

**Comment by wiztechinc** [comment]
Ditto, Scott. Some plugins just take more work than others and this particular one is essential for some of my clients. I think it would be a major selling point to have this functionality up front rather than as a additional plugin that someone might stumble over later.

**Comment by dakanji** [comment]
Everything that has been said is still easily achievable using officially supported modules.

Once something is a supported plugin, the work to support it needs to be in the core and this is what the core should be focused on. So shouldn't be a need to keep on patching for new releases.

With respect to concerned about people having to stumble upon such, they could, and should, be distributed with the program by default.

The advantage is that additional functionality, bug/security fixes can be rolled out in a modular fashion without having to wait for the whole thing to be updated or for threads with patches on the forum etc that someone hopefully stumbles upon.

In such cases, a new version of the affected module(s) is rolled out and job done. Particularly if they dial home and tell the user to upgrade.

Far easier for developers/maintainers and beneficial for users.

Anyway, just throwing out a thought for consideration. Point is not just on this one particular functionality but a general paradigm.

**Comment by lankeeyankee** [comment]
I think all of that is the end goal, but the plugin system needs to be working for this to be a reality. Is 1.5.7 going to wait for the plugin system? It's a big job and a turning point for zen cart as has been discussed before since it will break all existing modules. Right now, there is no way to add dual pricing without modifying core files. If the plugin system is maybe for 1.6.0 then I feel it should be added it now. Maybe @zcwilt can weigh in on where the owners are on this decision and a timeline.

**Comment by scottcwilson** [comment]
Possible 1.5.8 or future feature.

**Comment by torvista** [comment]
Hopefully I'll get my false teeth stuck into it before then!

**Comment by torvista** [comment]
Some months ago I looked at this and didn't like it.
I made my own from scratch using observers all the way. I still didn't like it as I found the manipulation of pricing when you get into the discounts to be a world of pain and so ended with a non-ideal/basic solution that is adequate for my needs but not for public consumption/release.
I still think the concept is valid for inclusion, but not until the pricing code is rewritten...is there a perceived need/plan for that?

**Comment by lat9** [comment]
> This mod is incredibly useful and would represent a significant new capability for Zen Cart.

One of the reasons that I'll put forward for its inclusion in the core is that it has many, many tentacles into core files, some of which haven't been massaged to add notifications over the past few years, e.g. attributes_controller.php and products_price_manager.php.

Given that it's an upgrade show-stopper for some sites and it _does_ have utility, I'd be for inclusion in the next non-patch-level release.  Noting that it would, IMO, need a configuration setting to enable/disable since its additional fields are just cluttering for sites that don't want/need the feature.

**Comment by proseLA** [comment]
personally i am not a fan of putting this into the core in its current state.

i think upgrade pain exists due to price function calls.  more functions creates more upgrade pain.

i am all for moving the `functions_prices` into a class file.  and then one can easily extend that class with the added functionality of a dual pricing.

**Comment by lat9** [comment]
It would _absolutely_ not go in "in its current form".  I'm looking to do some refactoring on the `Customer.php` class to accommodate the changes and heavily review the processing required to "make it right" (most likely including a configuration switch to enable/disable the processing so that sites that don't need it don't need to look at it).

**Comment by torvista** [comment]
>It would absolutely not go in "in its current form"

Agreed.

I tried to work with it/improve it but disagreed with so much of it I did my own solution.
The whole issue of trade/wholesale pricing should be evaluated as to whether it is a feature of value to enough users to be implemented "properly", perhaps by expanding the current customer groups.

Personally my trade prices are a percentage discount that varies per product (product-manufacturer) but are not volume-related as others would want.
I don't have different percentages per trade customer, as others would want.

It's complicated and once you get into the rabbithole of discount-pricing I don't think an add-on solution is viable that fits all permutations: too complex for a third-part to maintain altruistically.
So on that basis alone perhaps it should be built-in just because it's potentially so complex to add a comprehensive solution.

**Comment by lat9** [comment]
OK, here's my current thoughts/design documentation on the incorporation of this feature.

# Dual/Wholesale Pricing for zc200

Based on the `zen-cart-1.5.7` directory changes present in [this](https://github.com/lindasdd/zen-cart-dual-pricing) GitHub repository.

The *Wholesale Pricing* feature provides a means for a store-owner to define wholesale pricing tiers for products and each customer's record can have a pricing-tier designation.  When a customer has a wholesale-pricing account, products purchased by that customer can also be free from taxes (based on a configuration setting).

The existing "Dual Pricing" add-on provides these additional database fields:

| Table::Field                                 | Description                                                  |
| -------------------------------------------- | ------------------------------------------------------------ |
| customers::customers_whole                   | This field, if not '0', contains a numeric value indicating the customer's wholesale tier.  Its database definition is `VARCHAR( 4 ) DEFAULT '0' NOT NULL` which will be changed to `tinyint NOT NULL default 0`. |
| orders::is_wholesale                         | This field identifies whether (1) or not (0) the order was placed by a wholesale customer.  Note that the field value defaults to `NULL` for any orders placed prior to the incorporation of this change: `tinyint(1) DEFAULT NULL` and that this field was not previously defined in the add-on version. |
| products::products_price_w                   | This field's definition is `VARCHAR(150) DEFAULT '0' AFTER products_price` and is formatted as described below. |
| products_attributes::options_values_price_w  | This field's definition is `VARCHAR(150) DEFAULT '0' AFTER options_values_price` and is formatted as described below. |
| products_discount_quantity::discount_price_w | This field's definition is `VARCHAR(150) DEFAULT '0' AFTER discount_price` and is formatted as described below. |

Each of the wholesale price character fields is formatted as a `-` separated list of product-specific wholesale prices by pricing tiers.  If a store, for instance, has 4 fixed-pricing tiers, a wholesale pricing field might look like `1.00-2.00-3.00-4.00`.  If a store uses percent-off-pricing tiers, a wholesale pricing field might look like `1%-2%-3%-4%`.  Tiers can be a mixture of fixed- and percent-off-pricing.

***Notes***:

1. A customer cannot have both a *Discount Pricing Group* and a *Wholesale Pricing Tier*.  Whichever is set first prohibits the setting of the other.
2. *Sales* and *Specials* **do not apply** to wholesale customers.  Wholesale customers receive either their tier-specific wholesale price or the product's retail price.
3. If a product's wholesale pricing is _empty_ (i.e. the field is either an empty string or '0'), a product's retail pricing applies to all wholesale customers.
4. If a customer's wholesale pricing *tier* is _empty_ (i.e. the tier's value is an empty string or '0'), the wholesale customer's pricing is based on the next-lowest tier's pricing which also has a 'set' pricing.  For example:
   1. Given a wholesale pricing value of `0-1.00-2.00-3.00`, any customer with a wholesale pricing tier of `1` receives the product's retail pricing.
   2. Given a wholesale pricing value of `1.00--3.00`, any customer with a wholesale pricing tier of `2` receives the tier-1 pricing.  Any customer with a wholesale pricing tier `4` or greater receives the tier-3 pricing.

The integration of this add-on into the Zen Cart core requires changes for:

1. The `zc_install` process, adding the new fields if not already present and, if present, converting the `customers::customers_whole` field to a `tinyint` field.  An additional configuration setting (***My Store :: Wholesale Pricing***) will also be added to indicate whether wholesale pricing is enabled for the site (default 'false') and, if so, whether or not wholesale customers should be charged tax.
2. The admin's `Customers` tool, if wholesale pricing is enabled:
   1. Adds the data-entry of the integer-value *Wholesale Pricing Tier*.
   2. Disallows the addition of a *Wholesale Pricing Tier* when a customer also has a *Discount Pricing Group* and vice versa.
   3. Displays each customer's *Wholesale Pricing Tier* when the customers' listing is displayed.
3. The admin's `Catalog :: Categories/Products` tool, if wholesale pricing is enabled:
   1. Adds the data-entry of a product's *Wholesale Pricing Tiers*.
   2. Also copies a product's *Wholesale Pricing Tiers* from the original when a product is duplicated.
4. The admin's `Catalog :: Attributes Controller`, if wholesale pricing is enabled:
   1. Displays each attribute's *Wholesale Pricing* string when listing attributes assigned to a product.
   2. Displays a *Wholesale Pricing* text entry field for an attribute's data entry.
   3. ***Note***:  Unlike the add-on version used as a basis for these changes, this value is entered as a `-` separated string of absolute wholesale prices to be applied to a given attribute.  The add-on version forces the data entry to be a single, prefixed, floating point value.
5. The admin's `Catalog :: Products Price Manager`, if wholesale pricing is enabled:
   1. Adds a *Wholesale Discount* text entry field for each quantity-discount entry for the currently-displayed product.
6. `/includes/classes/shopping_cart.php`, if wholesale pricing is enabled:
   1. Class methods that perform pricing calculations (`calculate`, `attributes_price` and `get_products`) will be updated to use the wholesale product, attribute and discount pricing if the current customer has wholesale status.
7. `/includes/functions/functions_prices.php`
   1. `zen_get_products_special_price`.  Returns a product's base price if a wholesale customer is logged in; wholesale customers aren't eligible for sales or specials.
   2. `zen_get_products_base_price`.  Uses a product's wholesale pricing tier's price if a wholesale customer is logged in, unless overridden to force the return of the product's retail price (a newly-added function input).
   3. `zen_get_products_display_price`.  Formats the wholesale pricing display if a wholesale customer is logged in.
   4. `zen_get_attributes_final_price`.  Uses the wholesale pricing for any product attributes if a wholesale customer is logged in.
   5. `zen_get_products_discount_price_qty`.  Uses the wholesale pricing for any product quantity-discount if a wholesale customer is logged in.
   6. (**new**) `zen_get_retail_or_wholesale_price`.  This function is called whenever a product's price, attributes-price or quantity-discount price is needed for a pricing calculation.  It takes the associated price's base price and wholesale pricing tier as input and returns either the retail or wholesale price, depending on whether/not a wholesale customer is logged in.
   7. (**new**) `zen_get_products_retail_price`.  This function is an alias of `zen_get_products_base_price($product_id, true)`.
8. `/includes/functions/functions_products.php`
   1. `zen_products_lookup`.  ~~If the `products_price` field is requested, returns the value returned by `zen_get_retail_or_wholesale_price`.~~ Strike that; just modify to use `zen_get_product_details` to get the data in the cache for performance.
9. `/includes/functions/functions_taxes.php`
   1. `zen_get_tax_rate`.  Unconditionally returns a tax-rate of 0 if a wholesale customer is logged in and the site's configuration indicates that wholesale customers are always untaxed.
10. `/includes/modules/products_quantity_discounts.php`.
    1. Uses the `zen_get_retail_or_wholesale_price` function to retrieve the discount amount to use for the discount's pricing tier.
11. `/includes/classes/Customer.php`, adding static methods to be used by pricing/tax functions to determine wholesale customer status.
    1. `protected static function getWholesaleInfo()`.  This method contains a static variable (to cut down on database accesses) that identifies the session's current wholesale status.
    2. `public static function isWholesaleCustomer(): bool`.  This method returns a boolean indication identifying whether/not a wholesale customer is active in the session.
    3. `public static function isTaxExempt(): bool`.  This method returns a boolean indication identifying whether/not the customer is tax-exempt.  The method will include a notification enabling site- or plugin-specific overrides.
    4. `public static function getCustomerWholesaleTier(): int`.  This method returns an integer value identifying the wholesale tier to which the customer is authorized, returning `(int)0` if the customer is not a wholesale customer.
12. `/includes/classes/order.php`.
    1. The `create` method will be updated to set the `orders::is_wholesale` field to `(int)Customer::isWholesaleCustomer()`.

**Reply by torvista** [reply]
So the wholesale prices of a product may be a set of absolute reduced prices?...no contemplation of the various wholesale prices being various % discounts off the base price?
That is our policy and of every supplier we deal with: wholesale prices are always a percentage off.

**Reply by lat9** [reply]
That's the way the current add-on handles it.  I'd _seriously_ considered that addition and, with the brain-f@rt of having a common function to determine a wholesale customer's pricing, it's eminently do-able.

My only concern is the dreaded rounding errors that might result.

**Reply by lat9** [reply]
Documentation updated.

**Comment by lat9** [comment]
It looks like `/includes/modules/product_quantity_discounts.php` will also require changes; noting here so I don't forget!

Also will require changes to the Bootstrap template's `ZcaBootstrapObserver.php`, since it re-formats information for a product's displayed price.

**Comment by lat9** [comment]
Noting that the following PRs are prerequisites for this change:
- https://github.com/zencart/zencart/pull/6007
- https://github.com/zencart/zencart/pull/6011

**Comment by scottcwilson** [comment]
Should this be closed?
